File: kata-1/csharp/TimeToActDocumentAI.Tests/TestDataLoader.cs
using System.Text.Json;
using System.Text.Json.Nodes;

namespace TimeToActDocumentAI.Tests;

public record TestCase(
    string Name,
    string Description,
    string Input,
    JsonNode Expected
);

public record TestData(
    string Category,
    string Description,
    TestCase[] TestCases
);

public static class TestDataLoader
{
    private static readonly JsonSerializerOptions Options = new()
    {
        PropertyNameCaseInsensitive = true
    };

    public static TestData LoadTestData(string fileName)
    {
        var testDataPath = Path.Combine(
            Directory.GetCurrentDirectory(),
            "..", "..", "..", "..", "..", // Navigate up to kata-1 root
            "test-data",
            fileName
        );

        if (!File.Exists(testDataPath))
        {
            throw new FileNotFoundException($"Test data file not found: {testDataPath}");
        }

        var jsonContent = File.ReadAllText(testDataPath);
        var testData = JsonSerializer.Deserialize<TestData>(jsonContent, Options);
        
        if (testData == null)
        {
            throw new InvalidOperationException($"Failed to deserialize test data from {fileName}");
        }

        return testData;
    }

    public static IEnumerable<object[]> GetTestCases(string fileName)
    {
        var testData = LoadTestData(fileName);
        return testData.TestCases.Select(tc => new object[] { tc });
    }
}

File: katas/kata-1/fsharp/TimeToActParser.Tests/TestDataLoader.fs
module TimeToActParser.Tests.TestDataLoader

open System
open System.IO
open System.Text.Json
open System.Text.Json.Nodes

[<CLIMutable>]
type TestCase = {
    Name: string
    Description: string
    Input: string
    Expected: JsonNode
}

[<CLIMutable>]
type TestData = {
    Category: string
    Description: string
    TestCases: TestCase[]
}

let private options = JsonSerializerOptions()
options.PropertyNameCaseInsensitive <- true

let loadTestData (fileName: string) : TestData =
    let testDataPath = Path.Combine(
        Directory.GetCurrentDirectory(),
        "..", "..", "..", "..", "..", // Navigate up to kata-1 root
        "test-data",
        fileName
    )

    if not (File.Exists testDataPath) then
        failwith $"Test data file not found: {testDataPath}"

    let jsonContent = File.ReadAllText testDataPath
    let testData = JsonSerializer.Deserialize<TestData>(jsonContent, options)
    
    if obj.ReferenceEquals(testData, null) then
        failwith $"Failed to deserialize test data from {fileName}"
    else
        testData

let getTestCases (fileName: string) : TestCase[] =
    let testData = loadTestData fileName
    testData.TestCases